File: backend/src/AgentTeam/Services/WorkflowSchemaRepository.php
<?php

declare(strict_types=1);

namespace AgentTeam\Services;

use AgentTeam\Models\WorkflowSchema;
use PDO;

/**
 * Workflow Schema Repository
 *
 * CRUD for the workflow_schemas table — reusable JSON Schemas attached
 * to agent nodes for constrained decoding.
 */
class WorkflowSchemaRepository
{
    private PDO $db;

    public function __construct(PDO $db)
    {
        $this->db = $db;
    }

    /**
     * Find all schemas owned by a user.
     *
     * @return WorkflowSchema[]
     */
    public function findByUser(int $userId): array
    {
        $stmt = $this->db->prepare(
            "SELECT * FROM workflow_schemas WHERE user_id = ? ORDER BY name ASC"
        );
        $stmt->execute([$userId]);

        return array_map(
            fn($row) => new WorkflowSchema($row),
            $stmt->fetchAll(PDO::FETCH_ASSOC)
        );
    }

    public function findById(int $id): ?WorkflowSchema
    {
        $stmt = $this->db->prepare("SELECT * FROM workflow_schemas WHERE id = ?");
        $stmt->execute([$id]);
        $row = $stmt->fetch(PDO::FETCH_ASSOC);

        return $row ? new WorkflowSchema($row) : null;
    }

    public function findByName(int $userId, string $name): ?WorkflowSchema
    {
        $stmt = $this->db->prepare(
            "SELECT * FROM workflow_schemas WHERE user_id = ? AND name = ? LIMIT 1"
        );
        $stmt->execute([$userId, $name]);
        $row = $stmt->fetch(PDO::FETCH_ASSOC);

        return $row ? new WorkflowSchema($row) : null;
    }

    public function create(WorkflowSchema $schema): WorkflowSchema
    {
        $stmt = $this->db->prepare(
            "INSERT INTO workflow_schemas (user_id, name, description, schema_json, strict)
             VALUES (?, ?, ?, ?, ?)"
        );
        $stmt->execute([
            $schema->getUserId(),
            $schema->getName(),
            $schema->getDescription(),
            json_encode($schema->getSchemaJson(), JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE),
            $schema->isStrict() ? 1 : 0,
        ]);

        $schema->setId((int) $this->db->lastInsertId());
        return $schema;
    }

    public function update(WorkflowSchema $schema): WorkflowSchema
    {
        if (!$schema->getId()) {
            throw new \InvalidArgumentException('Cannot update schema without an ID');
        }

        $stmt = $this->db->prepare(
            "UPDATE workflow_schemas
             SET name = ?, description = ?, schema_json = ?, strict = ?
             WHERE id = ? AND user_id = ?"
        );
        $stmt->execute([
            $schema->getName(),
            $schema->getDescription(),
            json_encode($schema->getSchemaJson(), JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE),
            $schema->isStrict() ? 1 : 0,
            $schema->getId(),
            $schema->getUserId(),
        ]);

        return $schema;
    }

    public function delete(int $id, int $userId): bool
    {
        $stmt = $this->db->prepare(
            "DELETE FROM workflow_schemas WHERE id = ? AND user_id = ?"
        );
        $stmt->execute([$id, $userId]);
        return $stmt->rowCount() > 0;
    }
}

File: backend_python/app/agent_team/services/workflow_schema_repository.py
"""Port of backend/src/AgentTeam/Services/WorkflowSchemaRepository.php (112 lines).

CRUD for the workflow_schemas table — reusable JSON Schemas attached to
agent nodes for constrained decoding.
"""
from __future__ import annotations

from app.agent_team.models.workflow_schema import WorkflowSchema
from app.support.phpjson import dumps as php_json_dumps


class WorkflowSchemaRepository:
    def __init__(self, db):
        self.db = db

    def findByUser(self, user_id: int) -> list[WorkflowSchema]:
        """PHP 30-41."""
        rows = self.db.fetch_all(
            "SELECT * FROM workflow_schemas WHERE user_id = ? ORDER BY name ASC", [user_id]
        )
        return [WorkflowSchema(row) for row in rows]

    def findById(self, id: int) -> WorkflowSchema | None:
        """PHP 43-50."""
        row = self.db.fetch_one("SELECT * FROM workflow_schemas WHERE id = ?", [id])
        return WorkflowSchema(row) if row else None

    def findByName(self, user_id: int, name: str) -> WorkflowSchema | None:
        """PHP 52-61."""
        row = self.db.fetch_one(
            "SELECT * FROM workflow_schemas WHERE user_id = ? AND name = ? LIMIT 1",
            [user_id, name],
        )
        return WorkflowSchema(row) if row else None

    def create(self, schema: WorkflowSchema) -> WorkflowSchema:
        """PHP 63-79. json_encode with JSON_UNESCAPED_SLASHES|JSON_UNESCAPED_UNICODE."""
        new_id = self.db.insert(
            "INSERT INTO workflow_schemas (user_id, name, description, schema_json, strict)\n"
            "             VALUES (?, ?, ?, ?, ?)",
            [
                schema.getUserId(),
                schema.getName(),
                schema.getDescription(),
                php_json_dumps(schema.getSchemaJson()),
                1 if schema.isStrict() else 0,
            ],
        )

        schema.setId(new_id)
        return schema

    def update(self, schema: WorkflowSchema) -> WorkflowSchema:
        """PHP 81-102."""
        if not schema.getId():
            raise ValueError('Cannot update schema without an ID')

        self.db.execute(
            "UPDATE workflow_schemas\n"
            "             SET name = ?, description = ?, schema_json = ?, strict = ?\n"
            "             WHERE id = ? AND user_id = ?",
            [
                schema.getName(),
                schema.getDescription(),
                php_json_dumps(schema.getSchemaJson()),
                1 if schema.isStrict() else 0,
                schema.getId(),
                schema.getUserId(),
            ],
        )

        return schema

    def delete(self, id: int, user_id: int) -> bool:
        """PHP 104-111."""
        return self.db.execute(
            "DELETE FROM workflow_schemas WHERE id = ? AND user_id = ?", [id, user_id]
        ) > 0
